Don't report violations of VUID-vkCmdEndDebugUtilsLabelEXT-commandBuffer-01912

jimblandy opened the pull request:
As described in [Vulkan-ValidationLayers#5671], the validation layers don't understand debug ranges paired across different command buffers on the same queue, even though the Vulkan spec says:

> An application may open a debug label region in one command buffer and close it in another, or otherwise split debug label regions across multiple command buffers or multiple queue submissions. When viewed from the linear series of submissions to a single queue, the calls to vkCmdBeginDebugUtilsLabelEXT and vkCmdEndDebugUtilsLabelEXT must be matched and balanced.

Until this is fixed, wgpu should ignore this validation error to reduce noise in test runs.

Fixes #3733.

[Vulkan-ValidationLayers#5671]: https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5671

**Checklist**

- [X] Run `cargo clippy`.
- [X] Run `RUSTFLAGS=--cfg=web_sys_unstable_apis cargo clippy --target wasm32-unknown-unknown` if applicable.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

FrankenApps commented:
Here is the correct link to the Issue https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5671 in case anyone is interested.

jimblandy commented:
> Here is the correct link to the Issue [KhronosGroup/Vulkan-ValidationLayers#5671](https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5671) in case anyone is interested.

Fixed, thanks.

teoxoy commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -16,13 +16,30 @@ unsafe extern "system" fn debug_utils_messenger_callback(
+    const VUID_VKSWAPCHAINCREATEINFOKHR_IMAGEEXTENT_01274: i32 = 0x7cd0911d;
+    const VUID_VKCMDENDDEBUGUTILSLABELEXT_COMMANDBUFFER_01912: i32 = 0x56146426;
+
+    // Silence Vulkan Validation error " VUID-vkCmdEndDebugUtilsLabelEXT-commandBuffer-01912"
+    // This is a bug in the validation layer:
+    // https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5671
+    if cd.message_id_number == VUID_VKCMDENDDEBUGUTILSLABELEXT_COMMANDBUFFER_01912 {
+        return vk::FALSE;
+    }
```
If we can get the version of the validation layers, it would be better to conditionally ignore it.

jimblandy replied:
To narrow down the affected Mesa versions: it seems like the validation was introduced in KhronosGroup/Vulkan-ValidationLayers#5277, merged 2023-2-16.

jimblandy commented:
Layer properties in KhronosGroup/Vulkan-ValidationLayers@a90644606 (after introduction of 01912 test):

```
LayerProperties {
    layer_name: "VK_LAYER_KHRONOS_validation",
    spec_version: 4206832, // 1.3.240
    implementation_version: 1,
    description: "Khronos Validation Layer",
}
```

jimblandy commented:
Layer properties in https://github.com/KhronosGroup/Vulkan-ValidationLayers/commit/a54744d67 (current main):

```
JIMB: instance layer: LayerProperties {
    layer_name: "VK_LAYER_KHRONOS_validation",
    spec_version: 4206842, // 1.3.250
    implementation_version: 1,
    description: "Khronos Validation Layer",
}
```

jimblandy commented:
Okay, this version of the PR supplies the validation layer description and version to the debug utils messenger callback, and checks them to ensure that we're only filtering out errors from affected versions of `Vulkan-ValidationLayers`.

But I think it does violence to the API of `vulkan::Instance::from_raw`, so I want to look at it again tomorrow.

cwfitzgerald commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -236,48 +252,67 @@ impl super::Instance {
-            }
-            if log::max_level() >= log::LevelFilter::Warn {
-                severity |= vk::DebugUtilsMessageSeverityFlagsEXT::WARNING;
+        eprintln!(
```
Just putting a note on this println

jimblandy replied:
Yep, that's why it's a draft PR. :)

cwfitzgerald commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -236,48 +252,67 @@ impl super::Instance {
+            "JIMB: Instance::from_raw(debug_utils_user_data = {:?})",
+            debug_utils_user_data
+        );
+        let debug_utils = if let Some(debug_callback_user_data) = debug_utils_user_data {
+            if extensions.contains(&ext::DebugUtils::name()) {
```
Would be nice to have an info level print about why debug callbacks are disabled.

cwfitzgerald reviewed (changes requested):
Would be great to get this in as #3830 is now hitting it.

jimblandy committed:
Don't report violations of VUID-vkCmdEndDebugUtilsLabelEXT-commandBuffer-01912

As described in [Vulkan-ValidationLayers#5671], the validation layers don't understand debug ranges paired across different command buffers on the same queue, even though the Vulkan spec says:

> An application may open a debug label region in one command buffer and close it in another, or otherwise split debug label regions across multiple command buffers or multiple queue submissions. When viewed from the linear series of submissions to a single queue, the calls to vkCmdBeginDebugUtilsLabelEXT and vkCmdEndDebugUtilsLabelEXT must be matched and balanced.

Until this is fixed, wgpu should ignore this validation error to reduce noise in test runs.

Fixes #3733.

[Vulkan-ValidationLayers#5671]: https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/5671

cwfitzgerald reviewed (approved):
Nice!
